indispensable/src/commonMain/kotlin/at/asitplus/signum/indispensable/asn1/encoding/Asn1Decoding.kt

```diff
@@ -10,91 +10,162 @@ import at.asitplus.signum.indispensable.asn1.BERTags.T61_STRING
 import at.asitplus.signum.indispensable.asn1.BERTags.UNIVERSAL_STRING
 import at.asitplus.signum.indispensable.asn1.BERTags.UTF8_STRING
 import at.asitplus.signum.indispensable.asn1.BERTags.VISIBLE_STRING
+import at.asitplus.signum.indispensable.io.wrapInUnsafeSource
 import com.ionspin.kotlin.bignum.integer.BigInteger
 import com.ionspin.kotlin.bignum.integer.util.fromTwosComplementByteArray
 import kotlinx.datetime.Instant
+import kotlinx.io.Source
+import kotlinx.io.readByteArray
+import kotlinx.io.readUByte
 import kotlin.experimental.and
 
 
 /**
- * Parses the provided [input] into a single [Asn1Element]. Consumes all Bytes and throws if more than one Asn.1 Structure was found or trailing bytes were detected
- * @return the parsed [Asn1Element]
- *
- * @throws Asn1Exception on invalid input or if more than a single root structure was contained in the [input]
+ * Convenience wrapper around [parse], taking a [ByteArray] as [source]
+ * @see parse
  */
 @Throws(Asn1Exception::class)
-fun Asn1Element.Companion.parse(input: ByteIterator): Asn1Element = parseFirst(input).also {
-    if (input.hasNext()) throw Asn1StructuralException("Trailing bytes found after the first ASN.1 element")
-}
+@Deprecated(
+    "Use a ByteArray or (even better) a kotlinx.io Source as input when possible. This method copies all bytes from the input twice and is inefficient.",
+    ReplaceWith("source.readAsn1Element(); require(source.exhausted())")
+)
+fun Asn1Element.Companion.parse(input: ByteIterator): Asn1Element =
+    parse(mutableListOf<Byte>().also { while (input.hasNext()) it.add(input.nextByte()) }.toByteArray())
```

Comment:
i don't think there is

Reply:
write a two-line wrapper for `ByteIterator` -> `RawSource` then? it's better than copying everything

Reply:
same difference, because the first thing that will happen is, you make a call to `.buffered()` to make the source usable and then it tries to copy 8KiB into an internal buffer